app/domain/authentication/resource_restrictions/extract_resource_restrictions.rb

```diff
@@ -0,0 +1,120 @@
+require 'command_class'
+
+module Authentication
+  module ResourceRestrictions
+
+    ExtractResourceRestrictions = CommandClass.new(
+      dependencies: {
+        resource_restrictions_class: ResourceRestrictions,
+        role_class:                  ::Role,
+        resource_class:              ::Resource,
+        logger:                      Rails.logger
+      },
+      inputs:   %i(authenticator_name service_id host_name account)
+    ) do
+
+      def call
+        @logger.debug(LogMessages::Authentication::ExtractingRestrictionsFromResource.new(
+          @authenticator_name,
+          @host_name
+        ))
+
+        fetch_resource_annotations
+        extract_resource_restrictions_from_annotations
+        create_resource_restrictions_object
+
+        @logger.debug(LogMessages::Authentication::ExtractedResourceRestrictions.new(resource_restrictions.names))
+
+        resource_restrictions
+      end
+
+      private
+
+      def fetch_resource_annotations
+        resource_annotations
+      end
+
+      def resource_annotations
+        @resource_annotations ||=
+          host_resource.annotations.each_with_object({}) do |annotation, result|
+            annotation_values = annotation.values
+            value = annotation_values[:value]
+            next if value.blank?
+            result[annotation_values[:name]] = value
+          end
+      end
+
+      def host_resource
+        # Validate role exists, otherwise getting role annotations return empty hash.
+        role_id = @role_class.roleid_from_username(@account, @host_name)
+        host_resource = @resource_class[role_id]
+
+        raise Errors::Authentication::Security::RoleNotFound, role_id unless host_resource
```

Comment:
again, I think host_resource is not very presentative of a name if we involve users

Reply:
neither do `username`.
See my response above.

Reply:
what about `role_resource`?

Reply:
`role` and `resource` are two different db entities.
Took me a while to understand the difference between them.
I think combining them will only confuse more.

Reply:
renamed to `resource`.